### PR TITLE
enhance error handling with thiserror

### LINKED ISSUE #121: Use thiserror crate to simply errors across the repo
Errors like `DaemonError`, `RafsError` can be simplified with the abilities provided in `thiserror` crate.

For instance, display, from can be derived directly if using thiserror.

### PR DESCRIPTION
**Pull request by CormickKneey**
## Relevant Issue (if applicable)
close [issue](https://github.com/dragonflyoss/image-service/issues/121)

## Details
use thiserror in ApiError and RafsError. Also some changes in nydusError.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] make successfully

### CODE REVIEW

**Review by adam3q — COMMENTED** [review]
Thanks for your contribution!

BTW, you can use "git commit --amend -s" to sign this commit to pass the CI.

**Comment by CormickKneey** [comment]
It seems that there might be an issue with the aci pipeline. Could u help take a look or let me know if there is anything else I need to do?

**Comment by imeoer** [comment]
@CormickKneey The codecov CI should be fixed, let's move forward (the failed `codecov/patch` can be ignored).

**Review by adam3q — APPROVED** [review]
Thanks for your contribution!

File: api/Cargo.toml [hunk 1]
```diff
@@ -14,6 +14,7 @@ log = "0.4.8"
 serde_json = "1.0.53"
 toml = "0.5"
 
+thiserror = "1.0.30"
```
Comment (jiangliu) [inline api/Cargo.toml:17]: Please move it to line 15.

File: utils/src/trace.rs [hunk 2]
```diff
@@ -51,8 +52,9 @@ pub enum TraceClass {
 }
 }
 
-#[derive(Debug)]
+#[derive(Error, Debug)]
```
Comment (jiangliu) [inline utils/src/trace.rs:55]: how about using `#[derive(thiserror::Error, Debug)]`, it's more informative.

**Comment by imeoer** [comment]
Thanks for the PR! The target of using `thiserror & andhow` is to remove the `api/src/error.rs` file.

**Review by jiangliu — APPROVED** [review]

**Comment by CormickKneey** [comment]
> Thanks for the PR! The target of using `thiserror & andhow` is to remove the `api/src/error.rs` file. Any subsequent PRs will be submitted?

Ah, I didn't notice the `macro rules` in the `api/src/error.rs`  earlier. But I took a quick look, and I'm not quite ready to smoothly switch those content to `thiserror `yet. Maybe I'll complete this task in the next time. :imp:

**Pull request merged**